README.PERFORMANCE.DEPLOY.md

```diff
@@ -0,0 +1,262 @@
+
+```
+require('apps.intel.intel10g').num_descriptors = ring_buffer_size
+...
+config.app(c, "nic", require(device_info.driver).driver, {...})
+```
+
+The default of 512 seems too small, based on load test at IMIX line rate tests against lwaftr, 1024 or 2048 gave equally good results. Num_descriptors controls the Receive Descriptor Length on the Intel 82599 Controller, which determines the number of bytes allocated to the circular buffer. This value must be a multiple of 128 (the maximum cache line size). Since each descriptor is 16 bytes in length, the total number of receive descriptors is always a multiple of 8. In networking terms, this defines the ingress buffer size in packets (TODO: is this correct?). Larger ingress buffer can reduce packet loss while Snabb is busy handling other packets, but it will also increase latency for packets waiting in the queue to be picked up by Snabb.
+
+### Enable engine.busywait
+Defined in src/core/app.lua and enabled before calling engine.main() via
+
+```
+engine.busywait = true
+engine.main(...)
+```
+If true then the engine will poll for new data in a tight loop (100% CPU) instead of sleeping according to the Hz setting. This will reduce overall packet latency and increase throughput at the cost of utilizing the CPU hosting Snabb at 100%. 
+
+### Monitor ifInDiscards
+Snabb offers SNMP based ifInDiscards counters when SNMP is enabled. (TODO: need an easier way to expose these counters from the Intel register QPRDC).
+
+Enable SNMP in Snabb:
+
+```
+config.app(c, nic_id, require(device_info.driver).driver, 
+  {..., snmp = { directory = "/tmp", status_timer = 1 }, ... })
+```
+
+Then access ifInDiscards counter via od (the exact offset can be calculated from the file /tmp/0000:81:00.0.index):
+
+```
+od -j 305 -A none -N 4 -t u4 /tmp/0000\:81\:00.0
+      94543
+```
+Above example shows 94543 discarded packets at ingress on Port 0 since launching Snabb.
+
+## Qemu & Vhost-User
+[Vhost-User](http://www.virtualopensystems.com/en/solutions/guides/snabbswitch-qemu/) is used to connect Snabb with a high performance virtual interface attached to a Qemu based virtual machine.  This requires hugepages (explained further down) made available to Qemu:
+
+```
+cd <qemu-dir>/bin/x86_64-softmmu/
+qemu-system-x86_64 -enable-kvm -m 8000 -smp 2 \
+    -chardev socket,id=char0,path=./xe0.socket,server \
+    -netdev type=vhost-user,id=net0,chardev=char0 \
+    -device virtio-net-pci,netdev=net0,mac=02:cf:69:15:0b:00   \
+    -object memory-backend-file,id=mem,size=8000M,mem-path=/hugetlbfs,share=on \
+    -numa node,memdev=mem -mem-prealloc \
+    -realtime mlock=on  \
+    /path/to/img
+```
+
+The allocated memory must match the memory-backend-file size (example shows 8GB). While qemu will fail to boot if there isn't enough hugepages allocated, it is recommended to have some spare and note that the pages are split amongst the NUMA nodes. Check the paragraph on NUMA in this  document.
```

Comment:
Also it is worth noting here that the memory of the QEMU guest can not be smaller than the HugePage size and to be divisible to that size. Examples:

 - with 2MB huge page size - one can have 512MB guest RAM but can't have 511MB
 - with 1GB huge page size - one can have 1GB or 2GB RAM but can't have neither 512MB nor 1536MB (i.e. 1.5GB).